### PR TITLE
Add FunctionTool and allow Planner to use tools

### PR DESCRIPTION
**Pull request by philippjfr**
This PR does two things:

1. Adds `FunctionTool` which can wrap arbitrary functions and allow an LLM to call it by converting the arguments to a Pydantic model.
2. Gives the Planner the ability to use tools directly.

Note that in order to infer descriptions of the function arguments we add a dependency on `griffe`.

### CODE REVIEW

File: lumen/ai/agents.py [hunk 1]
```diff
@@ -118,28 +119,6 @@ def _steps_layout(self):
     def __panel__(self):
         return self.interface
 
-    async def _select_table(self, tables):
```
Comment (philippjfr) [inline lumen/ai/agents.py:121]: This seemed completely unused?!

Comment (ahuang11) [reply lumen/ai/agents.py:121]: It was for FuzzyTableMatching if none was found. There might still be a need of it if embeddings tool isn't able to find any table.

File: lumen/ai/agents.py [hunk 2]
```diff
@@ -225,6 +208,10 @@ async def respond(
 
 
 class ChatAgent(Agent):
+    """
+    ChatAgent provides general information about available data
+    and other topics  to the user.
+    """
```
Comment (ahuang11) [inline lumen/ai/agents.py:214]: Wondering why only SourceAgent/ChatAgent adds a docstring? Is it used?

File: lumen/ai/coordinator.py [hunk 3]
```diff
@@ -37,7 +39,7 @@ class ExecutionNode(param.Parameterized):
     Defines a node in an execution graph.
     """
 
-    agent = param.ClassSelector(class_=Agent)
+    agent_or_tool = param.ClassSelector(class_=(Agent, Tool))
```
Comment (ahuang11) [inline lumen/ai/coordinator.py:42]: Should this be generalized to Actor?

File: lumen/ai/coordinator.py [hunk 4]
```diff
@@ -477,8 +498,8 @@ async def _choose_agent(
     ):
         if agents is None:
             agents = self.agents
-        agents = [agent for agent in agents if await agent.applies(self._memory)]
-        agent_names = tuple(sagent.name[:-5] for sagent in agents)
+        agents = [agent for agent in agents if await agent.applies(self._memory)]+self._tools["__main__"]
```
Comment (ahuang11) [inline lumen/ai/coordinator.py:501]:
```suggestion
        agents = [agent for agent in agents if await agent.applies(self._memory)] + self._tools["__main__"]
```

File: lumen/ai/coordinator.py [hunk 5]
```diff
@@ -621,35 +643,39 @@ async def _make_plan(
...
         new_msg = dict(
-            role=user_msg['role'],
-            content=(
-                f"<user query>{user_msg['content']}</user query>"
-                f"{reasoning.chain_of_thought}"
-            )
+            role='assistant',
+            content=reasoning.chain_of_thought
         )
-        messages = messages[:-1] + [new_msg]
```
Comment (ahuang11) [inline lumen/ai/coordinator.py:630]: I vaguely remember `serialize` left an empty message.

File: lumen/ai/coordinator.py [hunk 6]
```diff
@@ -661,25 +687,48 @@ async def _resolve_plan(self, plan, agents, messages) -> tuple[list[ExecutionNod
...
+                    render_output=summarize_step.render_output
+                )
+            )
         plan.steps = steps
         return execution_graph, unmet_dependencies
 
     async def _compute_execution_graph(self, messages: list[Message], agents: dict[str, Agent]) -> list[ExecutionNode]:
-        agent_names = tuple(sagent.name[:-5] for sagent in agents.values())
+        tool_names = [tool.name for tool in self._tools["__main__"]]
+        agent_names = [sagent.name[:-5] for sagent in agents.values()]+tool_names
```
Comment (ahuang11) [inline lumen/ai/coordinator.py:721]:
```suggestion
        agent_names = [sagent.name[:-5] for sagent in agents.values()] + tool_names
```

File: lumen/ai/tools.py [hunk 7]
```diff
@@ -73,3 +70,71 @@ async def respond(self, messages: list[Message], **kwargs: dict[str, Any]) -> st
...
+
+    The function may also consume information in memory, e.g. the current
+    table or pipeline by declaring the requires parameter.
+
+    The function may also return context to add to the current working memory
+    by returning a dictionary and declaring the `provides` parameter. Any
+    keys listed in the provides parameter will be copied into working memory.
+    """
+
+    formatter = param.Parameter(default="{function}({args}) returned: {returns}")
```
Comment (ahuang11) [inline lumen/ai/tools.py:89]: Would be nice to have a `doc` here

File: lumen/ai/tools.py [hunk 8]
```diff
@@ -73,3 +70,71 @@ async def respond(self, messages: list[Message], **kwargs: dict[str, Any]) -> st
...
+
+    function = param.Callable(default=None, doc="""
```
Comment (ahuang11) [inline lumen/ai/tools.py:91]: Does this support async?

File: lumen/ai/agents.py (hunk 2 above)
Comment (philippjfr) [reply lumen/ai/agents.py:214]: Because I haven't finished writing docstrings.

Comment (philippjfr) [reply lumen/ai/agents.py:214]: Sorry, to be clear, yes it's used, as a docstring, all classes should have docstrings.

Comment (ahuang11) [reply lumen/ai/agents.py:214]: Can you clarify what Agent's docstrings are used for in FunctionTool?

File: lumen/ai/agents.py (hunk 1 above)
Comment (philippjfr) [reply lumen/ai/agents.py:121]: Okay, removing for now. Can re-add later.

File: lumen/ai/tools.py (hunk 8 above)
Comment (philippjfr) [reply lumen/ai/tools.py:93]: It does now. Was on my to-do list.

File: lumen/ai/agents.py (hunk 2 above)
Comment (philippjfr) [reply lumen/ai/agents.py:214]: They aren't. They are just docstrings.

Comment (ahuang11) [reply lumen/ai/agents.py:214]: Okay, I was confused about "to be clear, yes it's used"

**Review by ahuang11 — APPROVED** [review]

File: lumen/ai/coordinator.py [hunk 9]
```diff
@@ -705,17 +705,21 @@ async def _resolve_plan(self, plan, agents, messages) -> tuple[list[ExecutionNod
             )
             steps.append(step)
         last_node = execution_graph[-1]
-        if isinstance(last_node.actor, Tool) and 'AnalystAgent' in agents:
+        if isinstance(last_node.actor, Tool):
+            if "AnalysisAgent" in agents and all(r in provided for r in agents["AnalystAgent"].requires):
+                expert = "AnalysisAgent"
```
Comment (ahuang11) [inline lumen/ai/coordinator.py:710]: Analyst or Analysis?

**Review by ahuang11 — APPROVED** [review]

**Pull request merged**